fix: MCP authorization parameter implementation

omaryashraf5 opened the pull request:
# What does this PR do?
Adding a user-facing `authorization ` parameter to MCP tool definitions that allows users to explicitly configure credentials per MCP server, addressing GitHub Issue #4034 in a secure manner.


## Test Plan
tests/integration/responses/test_mcp_authentication.py

bbrowning commented:
Can you point me to where in the Responses API spec it has this `authentication` attribute? I only see `authorization` listed for MCP tools.

bbrowning commented:
> @bbrowning Thanks for your comment! Yes, I changed it to 'authorization' However, this static approach would only be helpful for MCP credentials that are hardcoded in tool definitions (long lived tokens). But its not ideal for cases where we need to have different mcp credentials per user.

I'm not sure I follow what you're saying. Every inference request passes in the tools available for that request. So, with every inference request, the client can pass in an updated token for any MCP servers that request references. And that means every user also passes in their own credentials. Or, am I misunderstanding how you intend this to work?

omaryashraf5 commented:
this approach is static within each individual response but dynamic across responses.

ashwinb commented on src/llama_stack/apis/agents/openai_responses.py:
```diff
@@ -479,6 +483,26 @@ class AllowedToolsFilter(BaseModel):
 
 
+@json_schema_type
+class MCPAuthorization(BaseModel):
```
we should adopt exactly the type specified here: https://platform.openai.com/docs/api-reference/responses/create#responses_create-tools-mcp_tool-authorization otherwise we are forking off our Responses API without a good reason. (This is beyond the fact that passing `{ username, password }` being not correct all.)

omaryashraf5 replied:
@ashwinb thanks for your comment, I was about to remove this class altogether.

mattf commented on src/llama_stack/apis/agents/openai_responses.py:
```diff
@@ -487,6 +487,7 @@ class OpenAIResponseInputToolMCP(BaseModel):
     :param server_label: Label to identify this MCP server
     :param server_url: URL endpoint of the MCP server
     :param headers: (Optional) HTTP headers to include when connecting to the server
+    :param authorization: (Optional) Bearer token authorization string (format: "Bearer <token>")
```
the format should be just the token.

what spec are you working from?

omaryashraf5 commented on src/llama_stack/providers/inline/agents/meta_reference/responses/tool_executor.py:
```diff
@@ -299,10 +299,18 @@ async def _execute_tool(
+                # Prepare headers with authorization from tool config
+                headers = dict(mcp_tool.headers or {})
+                if mcp_tool.authorization:
+                    # Don't override existing Authorization header (case-insensitive check)
```
sure thing, @ashwinb

ashwinb commented on src/llama_stack/apis/agents/openai_responses.py:
```diff
@@ -495,6 +496,9 @@ class OpenAIResponseInputToolMCP(BaseModel):
     server_label: str
     server_url: str
     headers: dict[str, Any] | None = None
+    # OAuth access token for MCP server authentication
```
don't add comments here which are not represented above in the comment block since this is part of the API

ashwinb reviewed (approved):
lgtm, will let others look over it for a bit before merging.

leseb commented on src/llama_stack/providers/inline/agents/meta_reference/responses/tool_executor.py:
```diff
@@ -299,12 +299,14 @@ async def _execute_tool(
-                        headers=mcp_tool.headers or {},
                         tool_name=function_name,
                         kwargs=tool_kwargs,
+                        headers=mcp_tool.headers,
```
Unrelated change? Why did the `headers` change?

ashwinb commented on src/llama_stack/providers/utils/tools/mcp.py:
```diff
@@ -27,6 +27,36 @@
+    if authorization:
+        # Check if Authorization header already exists (case-insensitive check)
+        existing_keys_lower = {k.lower() for k in headers.keys()}
+        if "authorization" in existing_keys_lower:
```
@mattf why is that a security risk, could you explain?

omaryashraf5 commented on src/llama_stack/providers/remote/tool_runtime/model_context_protocol/model_context_protocol.py:
```diff
@@ -56,7 +56,12 @@ async def invoke_tool(self, tool_name: str, kwargs: dict[str, Any]) -> ToolInvoc
 
         headers = await self.get_headers_from_request(endpoint)
-        return await invoke_mcp_tool(endpoint, headers, tool_name, kwargs)
+        return await invoke_mcp_tool(
```
`get_headers_from_request()` now extracts `Authorization` from provider data and returns it separately, so we can pass it via the dedicated `authorization` parameter instead of embedding it in the generic `headers` dict.

The signature of `invoke_mcp_tool()` changed - parameter order is now `(endpoint, tool_name, kwargs, headers=None, authorization=None)`. Using named parameters ensures correctness and clarity.

omaryashraf5 commented on src/llama_stack/providers/remote/tool_runtime/model_context_protocol/model_context_protocol.py:
```diff
@@ -45,7 +45,7 @@ async def list_runtime_tools(
             raise ValueError("mcp_endpoint is required")
         headers = await self.get_headers_from_request(mcp_endpoint.uri)
-        return await list_mcp_tools(mcp_endpoint.uri, headers)
+        return await list_mcp_tools(endpoint=mcp_endpoint.uri, headers=headers)
```
`list_mcp_tools`  function was updated and using named parameters ensures correctness and clarity.

mattf reviewed (changes requested):
this should be simply -
- [x] add `authz` field to `OpenAIResponseInputToolMCP`
- [x] pass `authz` to mcp client
- [x] sanitize response to not include `authz` field

we should not be providing multiple ways to provide the token.

we must not be passing a token for one service, e.g. inference, to another, e.g. mcp server.

ashwinb commented:
Btw see this comment from the Stainless bot now https://github.com/llamastack/llama-stack/pull/4052#issuecomment-3505240104 and see the associated python SDK diff https://github.com/stainless-sdks/llama-stack-client-python/compare/preview/base/add-mcp-authentication-param..preview/add-mcp-authentication-param -- looks all good.

omaryashraf5 commented:
## MCP Authentication Parameter Migration - Phase 1 & 2 Plan

## Phase 1: Add Authorization Parameter with Backward Compatibility (Implemented)

### What Was Done:
1. **API Changes** - Added `authorization: str | None = None` parameter to:
   - `/tool-runtime/list-tools` endpoint (Tool Runtime API)
   - `/tool-runtime/invoke` endpoint (Tool Runtime API)
   - Both documented as "OAuth access token for authenticating with the MCP server"

2. **Implementation Changes**:
   - Updated `list_runtime_tools()` and `invoke_tool()` in `model_context_protocol.py` to accept the new parameter
   - Implemented dual authorization support: accepts both old header-based AND new parameter-based auth for now but that will change after merging this PR and the stainless release.
   - Created `get_headers_from_request()` method to extract authorization from `X-LlamaStack-Provider-Data` header for backward compatibility
   - New parameter takes precedence when both are provided: `final_authorization = authorization or provider_auth`

3. **Security Layer**:
   - Added `prepare_mcp_headers()` utility in `mcp.py` that validates and prepares headers
   - **Strict validation**: Rejects if Authorization is found in the `headers` dict (security risk)
   - Enforces separation: users must use the dedicated `authorization` parameter instead
   - Automatically adds "Bearer " prefix when constructing final HTTP headers to MCP server

4. **Test Updates**:
   - Updated `api_recorder.py` to pass `authorization` parameter through patched tool methods
   - Integration tests continue using old header-based approach via `X-LlamaStack-Provider-Data` header
   - Added comments clarifying Phase 1 backward compatibility behavior

### Why Backward Compatibility Was Necessary:
- **Timing Issue**: New client SDK with `authorization` parameter doesn't exist yet
  - Waiting for Stainless to auto-generate
  - Current SDK only supports old header-based authentication
- **Test Dependencies**: Cannot update tests until new SDK is available
  - Tests currently use `extra_headers` with provider data: `{"mcp_headers": {uri: {"Authorization": "Bearer token"}}}`
  - New SDK will support clean parameter: `authorization="token"`

### Current Behavior (Phase 1):
```python
# Old approach (still works - backward compatible)
provider_data = {"mcp_headers": {uri: {"Authorization": f"Bearer {token}"}}}
auth_headers = {"X-LlamaStack-Provider-Data": json.dumps(provider_data)}
client.tool_runtime.list_tools(tool_group_id=id, extra_headers=auth_headers)

# New approach (already works!)
client.tool_runtime.list_tools(tool_group_id=id, authorization=token)

# If both provided, new parameter wins
client.tool_runtime.list_tools(
    tool_group_id=id, 
    authorization=token,        # ← This takes precedence
    extra_headers=auth_headers  # ← Ignored if above is provided
)
```
## Phase 2: Remove Backward Compatibility (after Stainless release)

ashwinb commented on tests/integration/responses/test_mcp_authentication.py:
```diff
@@ -0,0 +1,153 @@
+)
+
+
+def test_mcp_authorization_bearer(compat_client, text_model_id):
```
can you replace `compat_client` by `responses_client` and remove the LibraryClient skip -- things changed around this due to https://github.com/llamastack/llama-stack/pull/4146

ashwinb commented on tests/integration/responses/test_mcp_authentication.py:
```diff
@@ -0,0 +1,153 @@
+from .helpers import setup_mcp_tools
+
+# Skip these tests in replay mode until recordings are generated
+pytestmark = pytest.mark.skipif(
```
why is this there? can you not generate recordings?

ashwinb commented:
Re: https://github.com/llamastack/llama-stack/pull/4052#issuecomment-3529667043 

@omaryashraf5 that's just an insanely bloated comment made by an LLM with very little guidance and it makes me feel _less_ confident of your contributions, not _more_. Can you please avoid it? Using LLMs is of course become important now, but please guide them better so we as humans can review both the code as well as the commentary. Thank you!

ashwinb reviewed (approved):
lgtm, thanks for the iteration. as soon as this lands, I will land the client-sdk PR (which Stainless will generate automatically) and when that happens, we can move all the `tool-runtime` APIs to use the `authorization` parameter and kill ProviderData validator completely.
